### ISSUE TITLE
[Enhancement] Validate 32K Nemotron FSDP2 dynamic batching beyond the first update

### ISSUE THREAD

**Issue by xinyu-dev** [NVIDIA-NeMo/RL#4168]
## Goal

Match the completed PrimeRL Example 5 setup as closely as supported by NeMo RL + NeMo Gym: full-parameter, text-only Nemotron 3.5 Super GRPO using the same Harbor SimpleQA task split and grader.

| Setting | Completed PrimeRL Example 5 | NeMo attempt |
| --- | --- | --- |
| GPU placement | 16 training + 16 inference GPUs | Matched |
| Agent / task | Verifiers Bash + Harbor SimpleQA | Terminus2NemoGym + the same Harbor task split and grader |
| Parameter / optimizer CPU offload | Both off | Matched |
| Activation CPU offload | On, alongside checkpointing | No equivalent exposed in this AutoModel path; checkpointing on |
| Scheduling | Async; maximum policy age 1 | Async; maximum trajectory age 1; implementation differs |
| Batch / group size | 256 / 8 | Matched target; no dynamic replacement |
| Training parallelism | FSDP2 across 16 training GPUs: DP shard 16, DP replicate 1; TP / CP / EP = 1 / 1 / 1 | Matched |
| Precision | BF16 compute; FP32 stored parameters and optimization | FP32 stored parameters; initially FP32 moments, then BF16 retries |
| LR / betas / weight decay / grad norm cap | 1e-5 / 0.9, 0.999 / 0 / 1 | Matched |
| Context / output / maximum turns | 32K / 4K / 8 | Matched limits; actual conversations differ |
| Inference parallelism | 4 separate vLLM engines × TP4 = 16 GPUs; EP1, CUDA graphs, 0.85 memory fraction | Matched settings; different vLLM versions |
| Output-layer memory | Fused, chunked forward and backward | Full FP32 logits and training log-softmax |
| Validation | 16 before/after; temperature 0; interval 10 | Same settings; failed before final validation |
| Agent / verifier timeout | 600 / 120 seconds | Matched |
| Rollout concurrency | Adaptive 8–128, starting at 64 | Fixed maximum 128 |

FSDP2 shards parameters, gradients and optimizer state across the 16 training
ranks; each rank processes different trajectories. TP/CP/EP = 1 therefore still
uses distributed training, without tensor, context or expert parallelism.

## Environment

- NeMo RL: `6453bfd66bda1572654b679fb7a9194f3033b148` for the source paths below.
- AutoModel 0.7.0, submodule `6ae5ca23895d1e06c81cc6103732a20510833fda`; PyTorch 2.11.0+cu130; Transformers 5.15.1.
- Training workload: eight GB200 nodes, four GPUs/node; 16 training GPUs and 16 inference GPUs.
- NeMo Gym: `c3bac96314a59f28b896f597eb9845d175bb0252`; Terminus2NemoGym, maximum 8 turns, summarization disabled.

## Description

Native dynamic batching completed the first GRPO update but the second ran out of GPU memory in a MoE activation. We need a supported configuration or documented capacity limit for the 32K multi-turn workload above.

This is a **memory-capacity investigation**, not a confirmed dynamic-batching correctness bug. The batcher reduced padding and executed an update; its token budget does not bound every model activation or temporary allocation.

## Steps to reconstruct the tested workload

1. Use the environment above and the same Harbor SimpleQA split: 128 training questions and 16 held-out questions. Configure 32 prompts × 8 answers = 256 trajectories/update, 32K context and 4K output per call.
2. Keep FP32 stored parameters and Adam moments, BF16 compute, activation checkpointing on, parameter/optimizer CPU offload off, and TP/CP/EP all 1. Use separate 16-GPU training and inference pools.
3. Select native dynamic batching, with packing disabled:

   ```yaml
   policy:
     train_global_batch_size: 256
     max_total_sequence_length: 32768
     dynamic_batching:
       enabled: true
       train_mb_tokens: 32768
       logprob_mb_tokens: 32768
       sequence_length_round: 64
     sequence_packing:
       enabled: false
     offload_optimizer_for_logprob: false
     dtensor_cfg:
       enabled: true
       _v2: true
       activation_checkpointing: true
       cpu_offload: false
       tensor_parallel_size: 1
       context_parallel_size: 1
       expert_parallel_size: 1
       automodel_kwargs:
         backend:
           dispatcher: torch
           experts: torch_mm
   ```

4. Run native async Gym GRPO with maximum trajectory age 1 and a three-update target. Check memory after Adam state is initialized by the first update.

This is a reconstruction checklist, not a standalone deterministic OOM reproducer. The failed full configs and rollout inputs were removed; fresh live browsing produces different trajectories. The retained single-turn config is a successful baseline, not this Harbor test.

## Actual result

September 17, Slurm step `3086521.14`:

- Update 1 completed in **196.27 seconds**.
- Update 2 failed in AutoModel's Torch MoE activation (`relu2_deepep`), allocating a BF16 intermediate **before the vocabulary projection**.
- **3.44 GiB requested / 2.83 GiB free**; PyTorch allocated **172.34 GiB**, with **115.73 MiB reserved but unused**.
- **1/3 updates**, exit 1. Initial validation completed; final validation was not reached.

This is consistent with capacity pressure. Dynamic batching can reduce padding, but cannot remove real tokens or tensors retained for backward. Adam state is resident after update 1. Lowering the token budget below the padded length of a single trajectory reaches the [native budget assertion](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/distributed/batched_data_dict.py#L805-L817); the batcher does not split that trajectory automatically.

## Why PrimeRL's run fits

PrimeRL completed **3/3 updates plus final validation** using packing, activation CPU offload and a tiled output layer. These are additional memory controls beyond a token-budget batcher. Activation offload is particularly relevant to this pre-output-layer failure because it can free earlier saved tensors before the next MoE temporary allocation.

There was no identical-trajectory ablation, and model wrappers, kernels and harnesses differ. We cannot assign the memory gap to one feature or claim PrimeRL's batcher alone explains the result.

## Expected behavior / investigation request

- Identify the remaining peak on representative long trajectories after optimizer state is resident.
- Publish native settings that complete at least three updates at these workload limits, or document the additional memory/parallelism needed.
- Document that a token budget does not guarantee memory fit and must accommodate the longest padded trajectory.
- Keep loss/trajectory semantics unchanged; silently truncating or splitting trajectories is not the requested workaround.

Related memory capability: [activation CPU offload #4166](https://github.com/NVIDIA-NeMo/RL/issues/4166). Vocabulary tiling addresses a separate output-layer peak.

## Run references

- [Retained findings and reproduction guide (NVIDIA access)](https://gitlab-master.nvidia.com/xiyu/benchling-buidathon-super-nemorl/-/blob/xinyu-dev/super35-nemorl/examples/lyris/super35/README.md). Failed-run configs and raw logs were removed during cleanup; the guide preserves their recorded findings.
- [PrimeRL completed-run config](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/configs/resolved/trainer.json) and [results](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/summary.json) (repository access required).
